### PR TITLE
[FLINK-23734][table] Migrate ComponentFactory to the new Factory stack

### PR DESCRIPTION
**Pull request by Airblader**
## What is the purpose of the change

This migrate `ComponentFactory`, and thus `ParserFactory` and `PlannerFactory`, to the `Factory` stack and removes `ComponentFactory`. I did not maintain compatibility with the old stack since both interfaces are internal.

## Verifying this change

This change is already covered by existing tests, such as basically everything. :-)

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/delegation/PlannerFactory.java [hunk 1]
```diff
@@ -34,25 +34,82 @@
...
-    Planner create(
-            Map<String, String> properties,
-            Executor executor,
-            TableConfig tableConfig,
-            FunctionCatalog functionCatalog,
-            CatalogManager catalogManager);
+public interface PlannerFactory extends Factory {
+
+    default String factoryIdentifier() {
+        return getClass().getCanonicalName();
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/delegation/PlannerFactory.java:45]: can we align this with the executor factory?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/EnvironmentSettings.java [hunk 2]
```diff
@@ -225,9 +225,9 @@ public String getExecutor() {
 
     @Internal
     public Map<String, String> toPlannerProperties() {
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/EnvironmentSettings.java:227]: We don't need the `toPlannerProperties` anymore. The refactoring can be similar to FLINK-23482.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/factories/PlannerFactoryUtil.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import org.apache.flink.table.delegation.Planner;
+import org.apache.flink.table.delegation.PlannerFactory;
+import org.apache.flink.table.delegation.PlannerFactory.Context;
+import org.apache.flink.table.delegation.PlannerFactory.DefaultPlannerContext;
+
+import java.util.Map;
+
+/** Utility for discovering and instantiating {@link PlannerFactory}. */
+@Internal
+public class PlannerFactoryUtil {
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/factories/PlannerFactoryUtil.java:35]: Rename to `DelegationUtil` and add the methods for finding the executor as well. Currently, this is only a method in various `TableEnvironmentImpl`.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/delegation/BlinkPlannerFactory.java [hunk 4]
```diff
@@ -19,53 +19,43 @@
...
-import java.util.Arrays;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
+import java.util.Collections;
+import java.util.Set;
 
 /** Factory to construct a {@link BatchPlanner} or {@link StreamPlanner}. */
 @Internal
 public final class BlinkPlannerFactory implements PlannerFactory {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/delegation/BlinkPlannerFactory.java:32]: rename to `DefaultPlannerFactory`

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/delegation/BlinkPlannerFactory.java [hunk 5]
```diff
@@ -19,53 +19,43 @@
...
-        DescriptorProperties properties = new DescriptorProperties();
-        return properties.asMap();
-    }
-
-    @Override
-    public List<String> supportedProperties() {
-        return Arrays.asList(EnvironmentSettings.STREAMING_MODE, EnvironmentSettings.CLASS_NAME);
+    public Planner create(Context context) {
+        if (Boolean.parseBoolean(
+                context.getOptions().getOrDefault(EnvironmentSettings.STREAMING_MODE, "true"))) {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/delegation/BlinkPlannerFactory.java:47]: the runtime mode can be directly read from configuration or be its own flag in the context

**Review by twalthr — COMMENTED** [review]
Thanks @Airblader. I added some comments to align it more to the `ExecutorFactory` which works similar to the `PlannerFactory`. Maybe we can further simplify the methods required in `EnvironmentSettings`.

**Comment by Airblader** [comment]
@twalthr Pulling the executor factory logic into the util isn't possible because `ExecutorFactory` expects an implementation for `StreamExecutionEnvironment` not reflected in the interface, and this class is not accessible here. This seems sufficiently unrelated to warrant being solved separately.

File: flink-connectors/flink-connector-hive/src/main/resources/META-INF/services/org.apache.flink.table.factories.TableFactory [hunk 6]
```diff
@@ -13,4 +13,3 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
```
Comment (twalthr) [inline flink-connectors/flink-connector-hive/src/main/resources/META-INF/services/org.apache.flink.table.factories.TableFactory:14]: remove file entirely

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/EnvironmentSettings.java [hunk 7]
```diff
@@ -211,34 +207,23 @@ public boolean isBlinkPlanner() {
...
-        return properties;
+    public String getExecutor() {
+        return executor;
     }
 
     /** A builder for {@link EnvironmentSettings}. */
     public static class Builder {
-        private static final String BLINK_PLANNER_FACTORY =
-                "org.apache.flink.table.planner.delegation.BlinkPlannerFactory";
+        private final String planner = PlannerFactory.DEFAULT_PLANNER;
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/EnvironmentSettings.java:224]: `DEFAULT_PLANNER` -> `DEFAULT_IDENTIFIER`

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/delegation/PlannerBase.scala [hunk 8]
```diff
@@ -18,24 +18,34 @@
 
 package org.apache.flink.table.planner.delegation
 
+import java.lang.{Long => JLong}
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/delegation/PlannerBase.scala:21]: still wrong import order?

**Review by twalthr — COMMENTED** [review]
Thanks for the update @Airblader. I left my last comments. Should be good in the next iteration.

**Comment by Airblader** [comment]
@flinkbot run azure

**Pull request closed without merging**